lib/update/cluster/phases/kubernetes.go

```diff
@@ -378,4 +379,84 @@ func removeKubeletPermissions(client *kubeapi.Clientset) error {
 	return nil
 }
 
+func getInfluxDBNodename(client *kubeapi.Clientset, logger log.FieldLogger) (string, error) {
+	var labels map[string]string
+	deployment, err := getInfluxDBDeployment(client)
+	if deployment.Spec.Selector != nil {
+		labels = deployment.Spec.Selector.MatchLabels
+	}
+	replicaSets, err := rigging.CollectReplicaSets(deployment.Namespace, labels, logger, client, func(ref metav1.OwnerReference) bool {
+		return ref.Kind == rigging.KindDeployment && ref.UID == deployment.UID
+	})
+	if err != nil {
+		return "", trace.Wrap(rigging.ConvertError(err))
+	}
+
+	pods := make(map[string]v1.Pod, 0)
+	for _, replicaSet := range replicaSets {
+		podMap, err := rigging.CollectPods(replicaSet.Namespace, labels, logger, client, func(ref metav1.OwnerReference) bool {
+			return ref.Kind == rigging.KindReplicaSet && ref.UID == replicaSet.UID
+		})
+		if err != nil {
+			return "", trace.Wrap(rigging.ConvertError(err))
+		}
+		for nodename, pod := range podMap {
+			pods[nodename] = pod
+		}
+	}
+
+	if err != nil {
+		return "", trace.Wrap(rigging.ConvertError(err))
+	}
+	if len(pods) != 1 {
+		return "", trace.CompareFailed("unexpected number of InfluxDB pods running: %v, expected one pod", len(pods))
+	}
+	for nodename := range pods {
+		return nodename, nil
+	}
+	return "", nil
+}
+
+func getInfluxDBDeployment(client *kubeapi.Clientset) (*v1beta1.Deployment, error) {
+	deployment, err := client.ExtensionsV1beta1().Deployments(defaults.MonitoringNamespace).Get(defaults.InfluxDBDeploymentName, metav1.GetOptions{})
+	if err != nil {
+		return nil, trace.Wrap(rigging.ConvertError(err))
+	}
+
+	return deployment, nil
+}
+
+func upsertInfluxDBConfigMap(client *kubeapi.Clientset, logger log.FieldLogger, nodename string) error {
+	_, err := client.CoreV1().ConfigMaps(defaults.MonitoringNamespace).Create(&v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: defaults.MonitoringNamespace,
+			Name:      defaults.InfluxDBNodenameConfigMap,
+		},
+		Data: map[string]string{
+			defaults.InfluxDBConfigMapFieldName: nodename,
+		},
+	})
+	err = rigging.ConvertError(err)
+	if err != nil {
+		if trace.IsAlreadyExists(err) {
+			logger.WithField("configmap", defaults.InfluxDBNodenameConfigMap).Info("ConfigMap already exists, updating.")
+			err = updateInfluxDBConfigMap(client, nodename)
+		}
+	}
+	return trace.Wrap(err)
+}
+
+func updateInfluxDBConfigMap(client *kubeapi.Clientset, nodename string) error {
+	configMap, err := client.CoreV1().ConfigMaps(defaults.MonitoringNamespace).Get(defaults.InfluxDBNodenameConfigMap, metav1.GetOptions{})
```

Comment:
You don't really need to Get it here, you can just Update it if you got AlreadyExists on the Create call above.

We also have a function that does basically this: https://github.com/gravitational/gravity/blob/version/5.5.x/lib/ops/opsservice/monitoring.go#L214.

I would try to adopt that one instead: e.g. move it as exported function somewhere, say, to lib/utils/kubernetes.UpdateConfigMap and use here. Looking at it, it hardcodes the data map key, so you'd need to modify it a bit for example to accept the entire "data" map instead.